### ISSUE TITLE
Add generic loss function optimiser with simulator support

### ISSUE THREAD

**Issue by nikosavola** [gdsfactory/gdsfactory#1091]
**Is your feature request related to a problem? Please describe.**

Currently, gdsfactory appears to support SAX and MEEP optimisers. I suggest adding an example or support for a "generic optimisation loop", such that, a user would have to specify their geometry, search space, provide a loss function $\mathcal{L}$, methods for running the simulations and fetching their data.

In photonics, inverse design for individual elements appears popular. This approach is most likely very inefficient for that purpose. Instead, this framework should show how to combine simulation outputs and constraints from arbitrary sources for optimisation. Additionally, the choice of optimiser should be generic as well.

**Describe the solution you'd like**

I've found [Ray Tune](https://docs.ray.io/en/latest/tune/index.html) great for this purpose in not only hyperparamater selection but as a black-box optimiser for physics problems. Additionally, it already supports many other optimiser libraries, including [Ax](https://ax.dev/), [Optuna](https://optuna.org/), and [scikit-optimize](https://scikit-optimize.github.io/stable/). See [Search algorithms](https://docs.ray.io/en/latest/tune/api_docs/suggestion.html) for details.

I'll try providing a prototype example for this.

**Comment by joamatab** [comment]
yes Niko, a section of generic optimization sounds like a great idea

we can then apply it to different simulators and components in the docs

**Comment by nikosavola** [comment]
Closed after merging d851c46753b0390db6023269418f70605ece452b from #1137.
